algorithms/src/merkle_tree/merkle_tree.rs

```diff
@@ -41,10 +41,30 @@ pub struct MerkleTree<P: MerkleParameters> {
     parameters: Arc<P>,
 }
 
-impl<P: MerkleParameters> MerkleTree<P> {
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+impl<P: MerkleParameters + Send + Sync> MerkleTree<P> {
     pub const DEPTH: u8 = P::DEPTH as u8;
 
-    pub fn new<L: ToBytes, I: ExactSizeIterator<Item = L>>(parameters: Arc<P>, leaves: I) -> Result<Self, MerkleError> {
+    fn hash_row<L: ToBytes + Send + Sync>(
+        parameters: &P,
+        leaves: &[L],
+    ) -> Result<Vec<Vec<<<P as MerkleParameters>::H as CRH>::Output>>, MerkleError> {
+        let hash_input_size_in_bytes = (P::H::INPUT_SIZE_BITS / 8) * 2;
+        cfg_chunks!(leaves, 500)
```

Comment:
Can you clarify if `500` is a newly defined magical constant? We may want to parameterize this using a global const if so.

Reply:
Covered in above comment, it's experimentally found. I don't think this needs a global const.